.github/workflows/exercises.yml

```
@@ -0,0 +1,68 @@
+name: exercises
+on: [push, pull_request]
```

Note that this configuration will run duplicate builds on PR commits. I'd recommend this instead to avoid duplicate builds:
```suggestion
on:
  push:
    branches: [master]
  pull_request:
```

---

Thanks for taking a look.

My process on GitHub is generally:
1. Push a branch to my fork
1. Look at the CI run on that branch, and iterate as desired
1. Create a PR
1. Get code review and push more commits (or `fixup` previous commits) as desired

To my understanding, the suggested change has these pros and cons:
- Pro: it removes the duplicate `push`-triggered build that occurs when modifying a branch that already has a PR.
- Con: the `exercises.yml` workflow won't run on a branch (on my fork or on the main repo) **before** a PR is created for it.

Is that accurate? Is there something else?

Running CI without/before creating a PR is more important to me than removing the duplicate build. For example: I sometimes run CI on a branch without any intent to create a PR, or to find out if a PR is worthwhile.

I guess I want to run on a `push` **before** a PR is created, but not have a duplicate `push` build **after** the PR is created.

I believe that this:

```
on: 
  push:
  pull_request:
    types: [opened]
```

does what I want, except for one duplicate run when the PR is first created, which is fine by me.

Is this the most idiomatic approach? I don't really want to reach for some Action from the Marketplace.

---

> Is that accurate? Is there something else?

Correct.

> There's also this argument that they're not strictly duplicate builds:
> github.community/t/duplicate-checks-on-push-and-pull-request-simultaneous-event/18012/4

They are not exact duplicates, no. The `push` build targets the branch itself, while the `pull_request` build does a merge of the PR branch into the target branch, and targets that merged result for the build. So there is a possibility that the two have different results.

I'd defer to @SaschaMann to be sure, but I suspect `pull_request: {types: [opened]}` is not the right approach here.

> Running CI without/before creating a PR is more important to me than removing the duplicate build. For example: I sometimes run CI on a branch without any intent to create a PR, or to find out if a PR is worthwhile.

Then it seems reasonable to leave this as-is.

---

`types: [opened]` would only run on new PRs AFAIU.

You can add a condition to the job. If you *always* open PRs from a fork, it could look something like

```yaml
if: ${{ !(github.event_name == 'push' && github.event.pull_request.head.repo.full_name == github.repository) }}
```

(untested)

---

`types: [opened]` would only run on new PRs AFAIU.

You can add a condition to the job. If you *always* open PRs from a fork, it could look something like

```yaml
if: ${{ !(github.event_name == 'push' && 'exercism/nim' == github.repository) }}
```

(untested)

If you want it to build on branches of the main repo too, you'd have to add a check that checks for open PRs I think.

---

> Then it seems reasonable to leave this as-is.

Will do, thanks.

I think there are other justifications for `on: [push, pull_request]` too. Consider this case:
- A PR for branch `foo` has merge conflicts
- Therefore modifying branch `foo` does not trigger a `pull_request` build
- But there can be a good reason to modify branch `foo` before rebasing it on the main branch. For example, to apply a previously suggested change that itself must be modified when rebasing on the main branch.
- I want to see a CI run for the new changes before resolving the merge conflicts. To my understanding, that requires building on `push` more unconditionally.

> `types: [opened]` would only run on new PRs AFAIU.

Yeah. I believe it skips the `synchronize` builds when a PR's branch is modified - because the default `types` for `pull_request` are `opened`, `synchronize`, or `reopened` (docs [here](https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows#pull_request)). But the `pull_request` builds are more important than the `push` builds, so it's the wrong way round.

> You can add a condition to the job. If you _always_ open PRs from a fork, it could look something like

Nice. I do personally _always_ open PRs from a fork, which probably also makes the sometimes-superfluous `push` build a bit more acceptable (since it happens on the fork). But I'm not keen to write the general condition, which I think is something like: "run like `on: [push, pull_request]`, but don't run on `push` if there is an open PR that is free of merge conflicts"

It'd also be possible to skip the `push` build on `master` after merging a CI-passing PR that was strictly ahead of `master`, and show a little green checkmark that links to the previous successful run. But I prefer to keep things simple.